lib/translator/report_luns.cc

```diff
@@ -0,0 +1,125 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "report_luns.h"
+
+#ifdef __KERNEL__
+#include <linux/byteorder/generic.h>
+#else
+#include <netinet/in.h>
+#endif
+
+#include <byteswap.h>
+
+namespace translator {
+
+namespace {
+
+// Return the number of valid namespace IDs in a namespace ID list.
+uint32_t GetNsListLength(const nvme::IdentifyNamespaceList& ns_list) {
+  uint32_t size = 0;
+  for (uint32_t i = 0; i < nvme::kIdentifyNsListMaxLength; ++i) {
+    if (ns_list.ids[i] == 0) break;
+    ++size;
+  }
+  return size;
+}
+
+}  // namespace
+
+// Section 4.5
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+StatusCode ReportLunsToNvme(Span<const uint8_t> scsi_cmd,
+                            nvme::GenericQueueEntryCmd& identify_cmd,
+                            Allocation& allocation, uint32_t& alloc_len) {
+  // Cast scsi_cmd to ReportLunsCommand
+  scsi::ReportLunsCommand rl_cmd;
+  if (!ReadValue(scsi_cmd, rl_cmd)) {
+    DebugLog("Malformed ReportLuns command");
+    return StatusCode::kInvalidInput;
+  }
+  if (rl_cmd.select_report != scsi::SelectReport::kRestrictedMethods &&
+      rl_cmd.select_report != scsi::SelectReport::kWellKnown &&
+      rl_cmd.select_report != scsi::SelectReport::kAllLogical) {
+    DebugLog("Invalid report luns select report %u", rl_cmd.select_report);
+    return StatusCode::kInvalidInput;
+  }
+
+  // Assign allocation length for downstream use
+  alloc_len = ntohl(rl_cmd.alloc_length);
+
+  // Assign generic params to NVMe command
+  identify_cmd = {};
+  identify_cmd.opc = static_cast<uint8_t>(nvme::AdminOpcode::kIdentify);
+  identify_cmd.cdw[0] = 0x2;  // Set CNS to return namespace ID list
+
+  // Allocate prp & assign to command
+  if (allocation.SetPages(1, 0) == StatusCode::kFailure)
+    return StatusCode::kFailure;
+  identify_cmd.dptr.prp.prp1 = allocation.data_addr;
+  return StatusCode::kSuccess;
+}
+
+// Section 6.6
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+StatusCode ReportLunsToScsi(const nvme::GenericQueueEntryCmd& identify_cmd,
+                            Span<uint8_t> buffer) {
+  if (buffer.size() < sizeof(scsi::ReportLunsParamData)) {
+    DebugLog("Insufficient buffer size");
+    return StatusCode::kFailure;
+  }
+
+  // Get ns_list and lun count from NVMe dptr
+  uint8_t* ns_dptr = reinterpret_cast<uint8_t*>(identify_cmd.dptr.prp.prp1);
```

Comment:
is safe ptr meant to be used here?

Reply:
Unfortunately the buffer pointer `identify_cmd.dptr.prp.prp1` is of type uint64_t and SafePointerCastRead expects a span of type uint8_t. What's happening here is we are creating that span which is used in line 87